shared/connection.go

```diff
@@ -48,6 +68,74 @@ func MongoSession(uri string) *mgo.Session {
 	return session
 }
 
+func (opts MongoSessionOpts) configureDialInfoIfRequired(dialInfo *mgo.DialInfo) error {
+	if opts.TLSConnection == true {
+		config := &tls.Config{
+			InsecureSkipVerify: !opts.TLSHostnameValidation,
+		}
+		if len(opts.TLSCertificateFile) > 0 {
+			certificates, err := LoadKeyPairFrom(opts.TLSCertificateFile, opts.TLSPrivateKeyFile)
+			if err != nil {
+				return fmt.Errorf("Cannot load key pair from '%s' and '%s' to connect to server '%s'. Got: %v", opts.TLSCertificateFile, opts.TLSPrivateKeyFile, opts.URI, err)
+			}
+			config.Certificates = []tls.Certificate{certificates}
+		}
+		if len(opts.TLSCaFile) > 0 {
+			ca, err := LoadCaFrom(opts.TLSCaFile)
+			if err != nil {
+				return fmt.Errorf("Couldn't load client CAs from %s. Got: %s", opts.TLSCaFile, err)
+			}
+			config.RootCAs = ca
+		}
+		dialInfo.DialServer = func(addr *mgo.ServerAddr) (net.Conn, error) {
+			conn, err := tls.Dial("tcp", addr.String(), config)
+			if err != nil {
+				glog.Infof("Could not connect to %v. Got: %v", addr, err)
+				return nil, err
+			}
+			if config.InsecureSkipVerify {
+				err = enrichWithOwnChecks(conn, config)
+				if err != nil {
```

Comment:
Is this correct? Shouldn't this `return nil, err`? `enrichWithOwnChecks` returns closed connection on error. Returning error and closed connection might be misleading? This is not affecting anything so 👍